Add support for IAR version 8.x

deepikabhavnani opened the pull request:
Added support to IAR 8.x, with the reference of [release document](http://netstorage.iar.com/SuppDB/Public/UPDINFO/012120/arm/doc/infocenter/iccarm.ENU.html)

__aeabi_read_tp - Weak empty function is added to support the build. It should be implemented with TLS support in IAR Dlib.

Commit:
_MAX_LOCK define is moved to DLib_Threads.h

Commit:
__aeabi_read_tp function required for TLS in Dlib

Commit:
Internal members of std::FILE are not available from IAR8.x

Commit:
Use correct IAR toolchain macro

Commit:
Added new locking init requirement for v8.x

The application must call void __iar_Initlocks(void); before any
lock is used.

Commit:
Added function header for stl test

0xc0170 reviewed (commented):
What about testing? M0/M3/M4 with IAR all OK with these fixes (also backward compatible if possible. I assume yes as we are now using version < 8 macro in some changes here).

Commit:
__IAR_SYSTEMS_ICC__ macro had issues when 7.5 and 8.11 both workbench are installed

Commit:
Resolved compiler error: constructor deleted

IAR 8.x compiler throws error for emac_interface_t constructor
rtw_emac.cpp@220,0: [Pe1790]: the default constructor of "emac_interface"
cannot be referenced -- it is a deleted function

Error is because new object is created of structure with constant member.
Resolved it by using malloc instead new.

Commit:
Fix IAR8 : user-defined literal operator not found

Cellular example had build issues with IAR8, combination of define and string
as argument to send function resulted in above error. Typecasting to const char *
didn't help, hence replacing the define explicitly.

adbridge committed:
Remove string literal values and revert back to using CTX macro .

The previous fix to replace CTX with string literals was the wrong
solution. All that was actually required was to insert a space before
the macro.
